x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java

```diff
@@ -84,6 +85,14 @@ public TransportEsqlQueryAction(
             blockFactory
         );
         this.settings = settings;
+        /*
+         * This hacks the block factory into a shared place where
+         * it can be used for deserialization. We'd prefer a less
+         * strange way to do it, but Plugin doesn't give us access
+         * to BigArrays, which we need to build the BlockFactory
+         * up front.
+         */
+        blockFactoryHolder.blockFactory = blockFactory;
```

Comment:
Can we create a custom StreamInput, such as PlanStreamInput, and then create and pass a BlockFactory whenever we deserialize a Page? Currently, we only deserialize Pages in two places: ExchangeResponse and LookupResponse for enrich.

Reply:
++

Reply:
I'll have a look. I figured it'd amount to the same thing though because we have to figure out how to do that passing. But maybe it's later. I'll look.

Reply:
I did it! I do like this better.